Userland/Libraries/LibTLS/ClientHandshake.cpp

```diff
@@ -146,60 +143,70 @@ ssize_t TLSv12::handle_hello(ReadonlyBytes buffer, WritePacketStage& write_packe
+            return (i8)Error::NeedMoreData;
 
-            dbgln("Encountered extension {} with length {}", (u16)extension_type, extension_length);
-            // SNI
-            if (extension_type == HandshakeExtension::ServerName) {
-                u16 sni_host_length = AK::convert_between_host_and_network_endian(*(const u16*)buffer.offset_pointer(res + 3));
-                if (buffer.size() - res - 5 < sni_host_length) {
-                    dbgln("Not enough data for sni {} < {}", (buffer.size() - res - 5), sni_host_length);
+        if (extension_type == HandshakeExtension::ServerName) {
+            // RFC6066 section 3: SNI extension_data can be empty in the server hello
+            if (extension_length > 0) {
+                // ServerNameList total size
+                if (buffer.size() - res < 2)
                     return (i8)Error::NeedMoreData;
-                }
+                auto sni_name_list_bytes = AK::convert_between_host_and_network_endian(*(const u16*)buffer.offset_pointer(res += 2));
+                dbgln_if(TLS_DEBUG, "SNI: expecting ServerNameList of {} bytes", sni_name_list_bytes);
 
-                if (sni_host_length) {
-                    m_context.extensions.SNI = String { (const char*)buffer.offset_pointer(res + 5), sni_host_length };
-                    dbgln("server name indicator: {}", m_context.extensions.SNI);
-                }
-            } else if (extension_type == HandshakeExtension::ApplicationLayerProtocolNegotiation && m_context.alpn.size()) {
-                if (buffer.size() - res > 2) {
-                    auto alpn_length = AK::convert_between_host_and_network_endian(*(const u16*)buffer.offset_pointer(res));
-                    if (alpn_length && alpn_length <= extension_length - 2) {
-                        const u8* alpn = buffer.offset_pointer(res + 2);
-                        size_t alpn_position = 0;
-                        while (alpn_position < alpn_length) {
-                            u8 alpn_size = alpn[alpn_position++];
-                            if (alpn_size + alpn_position >= extension_length)
-                                break;
-                            String alpn_str { (const char*)alpn + alpn_position, alpn_length };
-                            if (alpn_size && m_context.alpn.contains_slow(alpn_str)) {
-                                m_context.negotiated_alpn = alpn_str;
-                                dbgln("negotiated alpn: {}", alpn_str);
-                                break;
-                            }
-                            alpn_position += alpn_length;
-                            if (!m_context.is_server) // server hello must contain one ALPN
-                                break;
+                // Exactly one ServerName should be present
+                if (buffer.size() - res < 3)
+                    return (i8)Error::NeedMoreData;
+                auto sni_name_type = (NameType) * (const u8*)buffer.offset_pointer(res++);
```

Comment:
This deref looks really weird...as if you're trying to multiply a pointer with a value called `NameType`...
is this clang-format's doing?

Reply:
Yeah I've seen `clang-format` do that sort of thing somewhere else too :(

Reply:
well...I guess one extra set of parenthesis is a small price to pay for more readable code (I had to think for a while whether something fishy was going on here, just because of _one_ space...)

@GMTA how does this look:
```diff
+ auto sni_name_type = (NameType) (*(const u8*)buffer.offset_pointer(res++));
- auto sni_name_type = (NameType) * (const u8*)buffer.offset_pointer(res++);
```

Reply:
Thanks @alimpfard, that looks more comprehensible - applied! I was checking to see if there were more appropriate C++-style casts I could apply but couldn't figure them out.

Reply:
Although `clang-format` really did not like that space between  `(NameType)` and `(*`...

Reply:
That's a bummer, but not a big problem 🤷‍♂️